File: SourceAFIS/Templates/Template.cs
using System;
using System.Collections.Generic;
using System.Text;
using SourceAFIS.General;

namespace SourceAFIS.Templates
{
    [Serializable]
    public sealed class Template : ICloneable
    {
        public enum MinutiaType : byte
        {
            Ending = 0,
            Bifurcation = 1,
            Other = 2
        }

        [Serializable]
        public struct Minutia
        {
            public readonly PointS Position;
            public readonly byte Direction;
            public readonly MinutiaType Type;

            public Minutia(TemplateBuilder.Minutia builderMinutia)
            {
                Position = new PointS(builderMinutia.Position);
                Direction = builderMinutia.Direction;
                switch (builderMinutia.Type)
                {
                    case TemplateBuilder.MinutiaType.Ending: Type = MinutiaType.Ending; break;
                    case TemplateBuilder.MinutiaType.Bifurcation: Type = MinutiaType.Bifurcation; break;
                    case TemplateBuilder.MinutiaType.Other: Type = MinutiaType.Other; break;
                    default: throw new ApplicationException();
                }
            }

            public TemplateBuilder.Minutia ToBuilderMinutia()
            {
                TemplateBuilder.Minutia builderMinutia = new TemplateBuilder.Minutia();
                builderMinutia.Position = Position;
                builderMinutia.Direction = Direction;
                switch (Type)
                {
                    case MinutiaType.Ending: builderMinutia.Type = TemplateBuilder.MinutiaType.Ending; break;
                    case MinutiaType.Bifurcation: builderMinutia.Type = TemplateBuilder.MinutiaType.Bifurcation; break;
                    case MinutiaType.Other: builderMinutia.Type = TemplateBuilder.MinutiaType.Other; break;
                    default: throw new ApplicationException();
                }
                return builderMinutia;
            }
        }

        public readonly int OriginalDpi;
        public readonly int OriginalWidth;
        public readonly int OriginalHeight;
        public readonly int StandardDpiWidth;
        public readonly int StandardDpiHeight;

        public readonly Minutia[] Minutiae;

        [NonSerialized]
        public object MatcherCache;

        public Template(TemplateBuilder builder)
        {
            OriginalDpi = builder.OriginalDpi;
            OriginalWidth = builder.OriginalWidth;
            OriginalHeight = builder.OriginalHeight;
            StandardDpiWidth = builder.StandardDpiWidth;
            StandardDpiHeight = builder.StandardDpiHeight;

            Minutiae = new Minutia[builder.Minutiae.Count];
            for (int i = 0; i < builder.Minutiae.Count; ++i)
                Minutiae[i] = new Minutia(builder.Minutiae[i]);
        }

        public TemplateBuilder ToTemplateBuilder()
        {
            TemplateBuilder builder = new TemplateBuilder();

            builder.OriginalDpi = OriginalDpi;
            builder.OriginalWidth = OriginalWidth;
            builder.OriginalHeight = OriginalHeight;

            foreach (Minutia minutia in Minutiae)
                builder.Minutiae.Add(minutia.ToBuilderMinutia());
            return builder;
        }

        public Template Clone()
        {
            return new Template(ToTemplateBuilder());
        }

        object ICloneable.Clone() { return Clone(); }
    }
}

File: java/sourceafis/src/sourceafis/templates/Template.java
package sourceafis.templates;

import java.io.Serializable;

  

@SuppressWarnings("serial")
public class Template implements Cloneable, Serializable {
 
	public int originalDpi;
	public int originalWidth;
	public int originalHeight;
	public int standardDpiWidth;
	public int standardDpiHeight;
	public Minutia[] minutiae;
	public Object matcherCache;
   /*
	public class Minutia {
		public PointS Position;
		public byte Direction;
		public MinutiaType Type;

		/*public Minutia(TemplateBuilder.Minutia builderMinutia) {
			Position = new PointS(builderMinutia.Position);
			Direction = builderMinutia.Direction;
			if (builderMinutia.Type == MinutiaType.Ending)
				Type = MinutiaType.Ending;
			else
				Type = MinutiaType.Bifurcation;
		}

		public TemplateBuilder.Minutia ToBuilderMinutia() {
			 TemplateBuilder.Minutia builderMinutia = new  TemplateBuilder.Minutia();// new
																							// TemplateBuilder.Minutia();
			builderMinutia.Position = Position.ToPoint();
			builderMinutia.Direction = Direction;
			if (Type == MinutiaType.Ending)
				builderMinutia.Type = MinutiaType.Ending;
			else
				builderMinutia.Type = MinutiaType.Bifurcation;
			return builderMinutia;
		} 
	}*/
	
	public Template(TemplateBuilder builder) {
		originalDpi = builder.originalDpi;
		originalWidth = builder.originalWidth;
		originalHeight = builder.originalHeight;
		standardDpiWidth = builder.getStandardDpiWidth();
		standardDpiHeight = builder.getStandardDpiHeight();

		minutiae = new Minutia[builder.minutiae.size()];
		for (int i = 0; i < builder.minutiae.size(); ++i)
			minutiae[i] = builder.minutiae.get(i);//new Minutia(builder.Minutiae.get(i));
	}

	public TemplateBuilder toTemplateBuilder() {
		TemplateBuilder builder = new TemplateBuilder();

		builder.originalDpi = originalDpi;
		builder.originalWidth = originalWidth;
		builder.originalHeight = originalHeight;
		for (Minutia minutia : minutiae) {
			//builder.Minutiae.add(minutia.ToBuilderMinutia());
			builder.minutiae.add(minutia);
		}
		return builder;
	}

	public Template clone() {
		return new Template(toTemplateBuilder());
	}

 
}
